src/main/java/com/google/sps/servlets/TagServlet.java

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.meltingpot.servlets;
+
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.CollectionReference;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.Query;
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.WriteResult;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.sps.meltingpot.data.DBObject;
+import com.google.sps.meltingpot.data.DBReferences;
+import com.google.sps.meltingpot.data.Recipe;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import javax.servlet.AsyncContext;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Handles recipe tag requests. */
+@WebServlet("/api/tag")
+public class TagServlet extends HttpServlet {
+  private Gson gson = new Gson();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String json = getRecipeList(request);
+
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  private String getRecipeList(HttpServletRequest request) {
+    Query query = DBReferences.tags();
+
+    ApiFuture<QuerySnapshot> querySnapshot = query.get();
+    ArrayList<Object> tagList = new ArrayList<>();
+
+    try {
+      for (DocumentSnapshot document : querySnapshot.get().getDocuments()) {
+        tagList.add(document.getData());
+      }
+      return gson.toJson(tagList);
+    } catch (InterruptedException e) {
+      System.out.println("Attempt to query tags raised exception: " + e);
+    } catch (ExecutionException e) {
```

Comment:
We've been doing this exact try catch block whenever we perform operations on the db, and it's repeated a lot in our codebase. Is there any way to have this in a single function or something like that?

Reply:
Yeah as long as we have a single query and we're getting every document that it matches, we could create a function that takes an argument of ArrayList<Object> and ApiFuture<QuerySnapshot>. Where would be a good place to put this function?

Reply:
i have a pr that puts this into DBUtils so once that gets merged you should pull master and use that

Reply:
Sounds good!